### PR TITLE
Get user history

### PR DESCRIPTION
**Pull request by ValentinChCloud**
Hello Björn,
I take this opportunity (#4) to propose a [function](https://github.com/ValentinChCloud/galaxy_ie_helpers/blob/master/galaxy_ie_helpers/__init__.py#L142)
Actually I use this to get the user history ( not a specific dataset). I think it could be useful in case where you want to get the list of all datasets and their names ( and choose only csv,txt etc..) , to display it to the user, and (this is the point) to display it on client side to have a dropdown button for example .
image

There is the R function to get the dataset and parse for only csv file_type.

```
gx_get_user_history <- function( history_id="None"){
    library('rPython')
    python.load("/usr/local/bin/get_user_history.py")
    x <- python.call("user_history",history_id)
    v<-list()
    l=length(x)
    for (y in 1:l) {
        if(is.null(x[[y]])) { 
            print("nothing to see")
        }else{
            if(x[[y]]$'extension' == 'csv'){
                name<-paste(x[[y]]$'hid',x[[y]]$'name')
                id<-unname(x[[y]]$'hid')
                v[[name]]<-id
            }
        }
     }

 return v
}
```
It's an shiny use case example : 

Once you get v as a named list, you can easely display it as show above. When the user click on the name of his dataset, in fact this is the hid wil be used to get the dataset.

### CODE REVIEW

File: galaxy_ie_helpers/__init__.py [hunk 1]
```diff
@@ -140,10 +140,26 @@ def get(dataset_id, history_id=None):
...
+       Get all visible dataset infos of user history.
+       Return a list of dict of each dataset.
+       The last element is always None
+    """ 
+    history_id = history_id or os.environ['HISTORY_ID']
+    gi = get_galaxy_connection(history_id=history_id, obj=False)
+    hc = HistoryClient(gi)
+    history = hc.show_history(history_id, visible=True, contents=True)
+    # I add a None element, to be sure the list has at least 2 elements
+    # It's neccessary, in the case case where the list give to R, it will
```
Comment (bgruening) [inline galaxy_ie_helpers/__init__.py]: s/case//

File: galaxy_ie_helpers/__init__.py [hunk 2]
```diff
@@ -140,10 +140,26 @@ def get(dataset_id, history_id=None):
...
+    # be interpreted as a list of list, and not only a list.
+    history.append(None)
```
Comment (bgruening) [inline galaxy_ie_helpers/__init__.py]: Can we hack around this on the R side? The None will be confusing I guess for other users of this function.

File: galaxy_ie_helpers/__init__.py [hunk 3]
```diff
@@ -140,10 +140,26 @@ def get(dataset_id, history_id=None):
...
+    return history
+
 
 if __name__ == '__main__':
     parser = argparse.ArgumentParser(description='Connect to Galaxy through the API')
-    parser.add_argument('--action',   help='Action to execute', choices=['get', 'put'])
+    parser.add_argument('--action',   help='Action to execute', choices=['get', 'put','get_history_user'])
```
Comment (bgruening) [inline galaxy_ie_helpers/__init__.py]: get_user_history?

**Review by bgruening — COMMENTED** [review]
Cool enhancement, I added a few comments and I think you also need to adjust the setup.py file to install the new binary.

**Comment by ValentinChCloud** [comment]
Yes, we can do this on the R side, but the function python.call of rPython library, will return : 
- a list in the case where there is only one dataset 
- a list of list in multiple datasets case

On the R side it could look like

```
library('rPython')
python.load("/usr/local/bin/get_user_history.py")
x <- python.call("user_history",history_id)
v<-list()

# If the first element is not a list, it's mean only one dataset

if(is.list(x[[1]]) == FALSE) {
    if(x$'extension' == 'csv'){
                name<-paste(x$'hid',x$'name')
                id<-unname(x$'hid')
                v[[name]]<-id
    }
}else{
    l=length(x)
    for (y in 1:l){
        if(x[[y]]$'extension' == 'csv'){
                name<-paste(x[[y]]$'hid',x[[y]]$'name')
                id<-unname(x[[y]]$'hid')
                v[[name]]<-id
            }
     }
}
```

This is a bit heavy, but it's imposed by the fact R doesn't handle dictionnary objet as Python (mentioned  [there](https://stackoverflow.com/questions/10678872/equivalent-of-a-python-dict-in-r)

I guess you right, I shouldn't have modified the python function to suit R problems,  the function has to be the most flexible possible.

**Comment by ValentinChCloud** [comment]
Hello @bgruening, I have noticed this PR is still open , may be it can be merged if it's ok for you?

**Comment by bgruening** [comment]
Sorry, forgot about that :(

**Comment by bgruening** [comment]
Thanks! @erasche is working on a new awesome concept of mounting in entire histories during container start, so this will also make your life easier we hope!

**Pull request merged**